node/lib/synthetic_gc.js

```diff
@@ -0,0 +1,231 @@
+            action: 'storeTrue',
+            defaultValue: false,
+        }
+    );
+    
+    return parser.parseArgs();
+}
+
+
+function* removeSyntheticRef(repo, commit) {
+    const refPath = SyntheticBranchUtil.getSyntheticBranchForCommit(commit);
+
+    console.log("Removing ref: " + refPath);
+
+    if (SIMULATION) {
+        return;
+    }
+    
+    return yield GitUtil.removeRemoteRef(repo, refPath);
+}
+
+// Go through the parents of 'commit' and remove synthetic reference recusively if they satisfy 'isDeletable'. 
+function* recursiveSyntheticRefRemoval(repo, commit, isDeletable, existingReferences) { 
+
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.instanceOf(commit, NodeGit.Commit);
+
+    if (commit.parentcount() === 0) {
+        return;
+    }
+
+    const parents = yield commit.getParents(commit.parentcount());
+    yield parents.map(function *(parent) {
+        if (parent in visited) {
+            return;
+        }
+        visited[parent] = 1;
+
+        if (isDeletable(parent) && existingReferences.includes(parent.sha())) {
+            yield removeSyntheticRef(repo, parent);
+        }
+        return yield *recursiveSyntheticRefRemoval(repo, parent, isDeletable, existingReferences);
+    });
+
+    return;
+}
+
+    // Clean up all redundant 
+function* cleanUpRedundant(repo, roots, predicate) {
+
+   for (let subName in roots) {
+       const subRepo = yield SubmoduleUtil.getRepo(repo, subName);
+
+       let existingReferences = yield getSyntheticRefs(subRepo);
+
+       for (let subCommit of roots[subName]) {
+           yield recursiveSyntheticRefRemoval(subRepo, subCommit, predicate, existingReferences); 
+       }
+   }
+}
+
+function* getSyntheticRefs(subRepo) {
+
+    const syntheticRefExtractor = function(value) {
+          if (value && value.includes(SYNTHETIC_BRANCH_BASE) ) {
+              return value.split("\t")[0];
+          }
+          return null;
+    }
+
+    let references = yield GitUtil.getRefs(subRepo);
+
+    return references.map(syntheticRefExtractor).filter(commit => commit);
+}
+
+    // Clean up all refs that satisfy 'isOldCommit', but not part of 'roots'.
+function* cleanUpOldRefs(repo, roots, isOldCommit) {
+
+   for (let subName in roots) {
+       const subRepo = yield SubmoduleUtil.getRepo(repo, subName);
+      
+       const reservedCommits = roots[subName];
+
+       let rawShaSubRoots = Array.from(reservedCommits).map(function(commit) {
+                            return commit.sha();
+                        });
+
+       let references = yield getSyntheticRefs(subRepo);
+
+       // filter out all the references from rootA
+       references = references.filter(ref => !rawShaSubRoots.includes(ref)); 
+
+
+       // filter out all the references younger than 6 months
+       let refsToDelete = [];
+       for (let ref in references) {
+           // could not figure out how to call generators withing filter properly, so cant be fancy
```

Comment:
Typically, you map to an array of callbacks, then call those callbacks in parallel by yielding the array, then you filter that.